catroid/src/main/java/org/catrobat/catroid/formulaeditor/datacontainer/DataContainer.java

```diff
@@ -24,350 +24,246 @@
-		name += " " + variableCounter;
-
-		while (getUserVariable(currentSprite, name) != null) {
-			variableCounter++;
-			name = originalName + " " + variableCounter;
+	public boolean containsProjectVariable(String name) {
+		for (UserVariable var : projectVariables) {
+			if (var.getName().equals(name)) {
+				return true;
+			}
 		}
-		name = originalName + " " + variableCounter;
-
-		return name;
+		return false;
 	}
 
-	public UserVariable getUserVariable(Sprite currentSprite, String name, UserBrick userBrick) {
-		UserVariable variableToReturn = spriteVariableBehaviour.get(currentSprite, name);
-		if (variableToReturn == null && userBrick != null) {
-			variableToReturn = userBrickVariableBehaviour.get(userBrick, name);
-		}
-		if (variableToReturn == null) {
-			variableToReturn = findProjectVariable(name);
+	public boolean containsVariableInAnySprite(String name) {
+		for (Sprite sprite : spriteVariables.keySet()) {
+			if (containsSpriteVariable(sprite, name)) {
+				return true;
+			}
 		}
-		return variableToReturn;
-	}
-
-	public boolean deleteUserVariableByName(String userVariableName) {
-		UserVariable variableToDelete = getUserVariable(getCurrentSprite(), userVariableName, getCurrentUserBrick());
-		return spriteVariableBehaviour.delete(getCurrentSprite(), userVariableName)
-				|| userBrickVariableBehaviour.delete(getCurrentUserBrick(), userVariableName)
-				|| getProjectVariables().remove(variableToDelete); //TODO
+		return false;
 	}
 
-	public boolean deleteUserListByName(String userListName) {
-		UserList listToDelete = getUserList(getCurrentSprite(), userListName);
-		return spriteListBehaviour.delete(getCurrentSprite(), userListName)
-				|| getProjectLists().remove(listToDelete); //TODO
+	public boolean containsSpriteVariable(Sprite sprite, String name) {
+		return getUserVariable(sprite, name) != null;
 	}
 
-	public void removeVariablesOfClones() {
-		spriteVariableBehaviour.removeCloneData();
-		spriteListBehaviour.removeCloneData();
+	public boolean containsProjectList(String name) {
+		for (UserList var : projectLists) {
+			if (var.getName().equals(name)) {
+				return true;
+			}
+		}
+		return false;
 	}
 
-	public void removeVariableListForSprite(Sprite sprite) {
-		spriteVariables.remove(sprite);
-		spriteListOfLists.remove(sprite);
-
-		for (UserBrick userBrick : sprite.getUserBrickList()) {
-			userBrickVariables.remove(userBrick);
+	public boolean containsListInAnySprite(String name) {
+		for (Sprite sprite : spriteListOfLists.keySet()) {
+			if (containsSpriteList(sprite, name)) {
+				return true;
+			}
 		}
+
+		return false;
 	}
 
-	public void resetAllDataObjects() {
-		resetAllUserLists();
-		resetAllUserVariables();
+	public boolean containsSpriteList(Sprite sprite, String name) {
+		return getUserList(sprite, name) != null;
 	}
 
-	private void resetAllUserVariables() {
-		resetUserVariables(getProjectVariables());
+	public UserVariable getUserVariable(Sprite sprite, String name) {
```

Comment:
Since `UserVariable` and `UserList` implement `UserData`, you could use generics to remove duplicate code for those classes nearly everywhere. e.g.

```java
private <T extends UserData> T getUserData(Sprite sprite, String name,
		Map<Sprite, List<T>> spriteMap, List<T> projectLists) { // TODO: Improve names :)
	if (spriteMap.containsKey(sprite)) {
		for (T list : spriteMap.get(sprite)) {
			if (list.getName().equals(name)) {
				return list;
			}
		}
	}

	for (T list : projectLists) {
		if (list.getName().equals(name)) {
			return list;
		}
	}

	return null;
}

public UserVariable getUserVariable(Sprite sprite, String name) {
	return getUserData(sprite, name, spriteVariables, projectVariables);
}
```

IMO generics are the way to go with methods related to `UserVariable` and `UserList`. That way the logic of both classes will always be in sync.

Reply:
Yes! I really like generics. I've been thinking about it. Have you tried it out? For the first version you proposed you'll need some interface (e.g. UserData, as you proposed) for the ```getName(..)``` method and sometimes casting stuff (especially when it comes to ```List<T>```object and more so when it comes to ```List<List<T>>```objects) becomes problematic. Java sometimes behaves very peculiar when it comes to type security. This is basically the only reason why I haven't changed it yet.

I think I'll have another look at it 😉.

Reply:
> Have you tried it out?

I only checked for errors in the IDE so far :smile:
The only code piece that can't be shared is the instantiation of `UserVariable` and `UserList`, I guess (excluding `resetData`  as mentioned below). That's why I suggested the classes `UserDataContainer`, `UserVariableContainer`, … as an alternative which can also handle it more elegantly.
However, imo the preferable solution (generic classes or helper methods) all depends on how many methods can be reused for the `UserBrick` stuff. I still haven't looked at that :grin: Maybe the `UserBrick`s and the try to create a class hierarchy (behaviour classes) which suites all is one reason why the code got so ugly…

> you'll need some interface (e.g. UserData, as you proposed)…

The interface isn't my idea, both `UserVariable` and `UserList` already implement [`UserData.java`](https://github.com/Catrobat/Catroid/blob/develop/catroid/src/main/java/org/catrobat/catroid/formulaeditor/UserData.java). :wink: I would've named it `Nameable` though because it can be reused for all other objects like `Project`, `Sprite`, `Look`, … and then be used in utils methods. However, that's another topic.

> …and sometimes casting stuff…

I wouldn't expect a single cast (yet). I hope it stays that way and that I also didn't oversee a case :crossed_fingers: